### PR TITLE
Bug Fix: Refactor class `Entity`, `Environment`, `Resource`

### PR DESCRIPTION
**Pull request by raoyitao**
* Refactor `start/start_in_pool/stop/stop_in_pool` of `Environment`
  class to make code clean.
* Fix a bug that int `Environment.start` the `_wait_started` is called
  twice if a resource's `async_start` attribute is False.
* Fix a bug that in `Environment.start_in_pool` a resource will be
  started even its `auto_start` attribute is False.
* No need to maintain 2 copies of `start/stop` methods of `Resource`
  and its derive class `Driver`, the only difference is that there are
  customized `pre_start/post_start/pre_stop/post_stop` to be called.
  The `Environment` class works for `Driver` but it should be adapted
  to `Resource` (e.g. `resource.wait` is called twice unexpectedly).
* While start or stop resources, should check if it's already aborted,
  or the status transition will not be correct and raise exception.
* If exception raised during `stop`, the resource status should also
  be updated to avoid possible issues (In interactive base module we
  have some code to deal with the status that could be removed).
* Move `post_start` and `post_stop` methods out of wait handlers, it's
  confusing that `pre_start` and `post_start` are called in `start`,
  while the others are called in a protected methods, hard to maintain.
* `Worker`s are maintained by `Pool`, so no need to change status to
  `STARTED` in its `staring` method, in this time worker is not fully
   initialized and the heartbeat check from pool instance might fail.
* `_log_exception` always places error messages into `start_exceptions`
  but for `stop_in_pool` error messages should go to `stop_exceptions`.
* In debug log use '%s' format rather than '%r' because unique name is
  fairly enough, better than "CLASS_NAME <object at 0x00...>".
* Just like `start` and `stop`, it's better log basic debug info in
  `abort` method rather than in `aborting` method everywhere.
* Fix wrong log in `_abort_entity` method.
* `__eq__` method is already added for `ResourceStatus` so we can use
  `res.status == ST` instead of `res.status.tag == ST` everywhere.
* Remove lint error on try...finally... statements.
* Update testcases.

## Checklist:
- [x] Test
- [ ] Example (both test_plan.py and .rst)
- [ ] Documentation (API)
- [ ] News fragment present for release notes
- [ ] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [ ] For new cmdline arg: documentation

### CODE REVIEW

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: testplan/common/entity/base.py [hunk 1]
```diff
@@ -253,18 +236,17 @@ def stop(self, is_reversed=False):
                 )
                 resource.logger.error(msg)
                 self.stop_exceptions[resource] = msg
+                # Resource status should be STOPPED even it failed to stop
+                resource.status.change(resource.STATUS.STOPPED)
+            else:
+                if resource.cfg.async_start:
```
Comment (kn-ms) [inline testplan/common/entity/base.py:242]: Relay on resource.cfg sounds something that is internal to an Entity. Like an entity can decide it's asyncness not only based on the cfg.async value, would be better to rely on a property of the entity instead of the cfg of the enity. Though this might be a bigger change and itmay go into another PR

File: testplan/testing/multitest/driver/base.py [hunk 2]
```diff
@@ -207,16 +205,10 @@ def stopping(self):
     def _wait_started(self, timeout=None):
         self.started_check(timeout=timeout)
         self.status.change(self.STATUS.STARTED)
-        if self.cfg.post_start:
-            self.cfg.post_start(self)
-        self.post_start()
```
[the remark this answers is not in the document]
Comment (raoyitao) [reply testplan/testing/multitest/driver/base.py:212]: If `async_start` is False, the entity itself wait status changed to STARTED and then `post_start` is called, if `async_start` is True, its parent calls `wait` to check if the resource status changed and then calls `post_start` (`post_stop` works similarly). I just move `post_start` and `post_stop` out of a wait handler (which is a protected method), but did not modify the process, in `async_start` mode, the wait handler is also called by parent.

File: testplan/testing/multitest/driver/base.py [hunk 3]
```diff
@@ -161,28 +161,26 @@ def uid(self):
...
 
-    def pre_start(self):
-        """Steps to be executed right before driver starts."""
-        self.make_runpath_dirs()
+    def post_start(self):
+        """Steps to be executed right after resource is started."""
+        if self.cfg.post_start:
+            self.cfg.post_start(self)
+
+    def pre_stop(self):
```
[the remark this answers is not in the document]
Comment (raoyitao) [reply testplan/testing/multitest/driver/base.py:175]: I think no user wrote a class that inherits `Resource` or `Executor`, so this change will not affect them. However, User might write their own drivers that inherits `Driver` or `App`, then it depends. As a general principle, when a class derives from a base class, some resource handling function, like `setup`, `teardown`, `pre_XXX`, `post_YYY` should also be called, like this:

**def initialize_work(self):**
... **super(MyClass, self).initialize_work()**  # add this line, unless we clearly know we don't need it
... # code of derived class

You can have a look at our drivers, e.g. `KDB`, `CPSServer`, `ZookeeperStandalone`, they all follow the rules. But I'm not sure if our users are aware of this when they wrote their own drivers, it they took ours as a reference they will did it.

File: testplan/common/entity/base.py (hunk 1 above)
Comment (raoyitao) [reply testplan/common/entity/base.py:242]: I saw that there are several places `cfg.async_start` and `cfg.auto_start` occur in our code, not much change needed, so I add 2 properties into `Resource` class, then directly call `resource.async_start` and `resource.auto_start`.

File: testplan/common/entity/base.py [hunk 4]
```diff
@@ -282,22 +264,47 @@ def stop_in_pool(self, pool, is_reversed=False):
...
+
+    def _log_exception(self, resource, func, exception_record):
+        """
+        Decorator for logging an exception at resource and environment level.
+
+        :param resource: resource to log the exception with
+        :type resource: :py:class:`~testplan.common.entity.base.Resource`
+        :param func: function to catch exception for
+        :type func: ``Callable``
+        :param exception_record: A dictionary that maps resource name to
```
[the remark this answers is not in the document]
Comment (raoyitao) [reply testplan/common/entity/base.py:291]: yeah I think it's enough to just match the name, but I saw somewhere the resource is fetch from `exceptions.keys()` and its methods (e.g. `abort`, `fetch_error_log`) are called. So for this time I will not change it to avoid unwanted CI issue.

File: testplan/common/entity/base.py [hunk 5]
```diff
@@ -253,18 +236,17 @@ def stop(self, is_reversed=False):
                 )
                 resource.logger.error(msg)
                 self.stop_exceptions[resource] = msg
+                # Resource status should be STOPPED even it failed to stop
+                resource.status.change(resource.STATUS.STOPPED)
```
[the remark this answers is not in the document]
Comment (kn-ms) [reply testplan/common/entity/base.py]: resource.wait(resource.STATUS.STOPPED) <---- this waiting the status to become something but the above code is setting the status of the resource from outside

File: testplan/common/entity/base.py [hunk 6]
```diff
@@ -253,18 +236,17 @@ def stop(self, is_reversed=False):
                 )
                 resource.logger.error(msg)
                 self.stop_exceptions[resource] = msg
+                # Resource status should be STOPPED even it failed to stop
+                resource.wait(resource.STATUS.STOPPED)
```
Comment (Pyifan) [inline testplan/common/entity/base.py:240]: I saw this has been changed to wait(), which I do not think is right. Because stop() has exceptioned so no one is changing the status to STOPPED, so wait here might just hang？

Also, if resource failed to stop, what status should it have? Maybe we don't bother to change the status here in the first place. Because it's meant to be kind of undefined, it's not really STOPPED right？

**Review by Pyifan — APPROVED** [review]

**Review by Pyifan — CHANGES_REQUESTED** [review]
unapprove

File: testplan/common/entity/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (raoyitao) [reply testplan/common/entity/base.py]: Maybe this `wait` method is confusing, it means call the handler for that `status`, here, `resource.wait` will call `_wait_stopped`, it simply changes the `status` to `STOPPED`, or in derived class it wait until `status` changed to `STOPPED`.

File: testplan/common/entity/base.py (hunk 6 above)
Comment (raoyitao) [reply testplan/common/entity/base.py:240]: Kris had a comment at the same place, but he is concerning the private access of resource. I left a comment there for this issue.

**Review by Pyifan — APPROVED** [review]

File: testplan/testing/multitest/driver/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelsimonelli) [reply testplan/testing/multitest/driver/base.py:175]: @raoyitao is there an example implementation of  KDB driver?

Curious about implementation for running integration tests on a dynamic environment.

At first glance, it makes sense to inherent TCP client/server.

The server driver would start the kdb binary/instance on an ephemeral port,  load any configured scripts/options, can effectively recreate a micro environment (e.g. tickplant, RDB, HDB)

Is a separate kdb client driver absolutely necessary at this point? 

Could we not embed a q/python IPC client/interface directly into the server driver to send commands/queries?

Or does testplan expect/prefer to implement and communicate with driver client?
